File: solutions/swift/freelancer-rates/Sources/FreelancerRates/FreelancerRates.swift
let workHoursPerDay: Double = 8
let billableDaysPerMonth: Double = 22

// Calculate the daily rate given an hourly rate
func dailyRateFrom(hourlyRate: Int) -> Double {
  return Double(hourlyRate) * workHoursPerDay
}

// Calculate the monthly rate, given an hourly rate and a discount
func monthlyRateFrom(hourlyRate: Int, withDiscount discount: Double) -> Double {
  let discountAdjustment: Double = 1.0 - discount / 100
  let monthlyRate: Double = dailyRateFrom(hourlyRate: hourlyRate) * billableDaysPerMonth
  return (monthlyRate * discountAdjustment).rounded()
}

// Calculate the number of workdays given a budget, hourly rate and discount
func workdaysIn(budget: Double, hourlyRate: Int, withDiscount discount: Double) -> Double {
  let discountAdjustment: Double = 1.0 - discount / 100
  let dailyRate: Double = dailyRateFrom(hourlyRate: hourlyRate)
  return (budget / (dailyRate * discountAdjustment)).rounded(.down)
}

File: solutions/javascript/freelancer-rates/freelancer-rates.js
const workHoursPerDay = 8;
const billableDaysPerMonth = 22;
/**
 * The day rate, given a rate per hour
 *
 * @param {number} ratePerHour
 * @returns {number} the rate per day
 */
export function dayRate(ratePerHour) {
  return ratePerHour * workHoursPerDay;
}

/**
 * Calculates the number of days in a budget, rounded down
 *
 * @param {number} budget: the total budget
 * @param {number} ratePerHour: the rate per hour
 * @returns {number} the number of days
 */
export function daysInBudget(budget, ratePerHour) {
  return Math.floor(budget / dayRate(ratePerHour));
}

/**
 * The month rate, given a rate per hour and monthly discount
 *
 * @param {number} ratePerHour
 * @param {number} discount: for example 20% written as 0.2
 * @returns {number} the rate per day
 */
function monthRate(ratePerHour, discount) {
  return dayRate(ratePerHour) * billableDaysPerMonth * (1 - discount);
}

/**
 * Calculates the discounted rate for large projects, rounded up
 *
 * @param {number} ratePerHour
 * @param {number} numDays: number of days the project spans
 * @param {number} discount: for example 20% written as 0.2
 * @returns {number} the rounded up discounted rate
 */
export function priceWithMonthlyDiscount(ratePerHour, numDays, discount) {
  const numMonths = Math.floor(numDays / billableDaysPerMonth);
  const priceForMonths = monthRate(ratePerHour, discount) * numMonths;
  const remainingDays = numDays - numMonths * billableDaysPerMonth;  
  return Math.ceil(priceForMonths + remainingDays * dayRate(ratePerHour));
}

